lib/dataTransformer.ts

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import * as rhea from "rhea";
+import * as debugModule from "debug";
+
+const isBuffer = require("is-buffer");
+const debug = debugModule("azure:event-hubs:datatransformer");
+
+export interface DataTransformer {
+  /**
+   * @property {Function} encode A function that takes the body property from an EventData object
+   * and returns an encoded body (some form of AMQP type).
+   */
+  encode: (body: any) => any;
+  /**
+   * @property {Function} decode A function that takes the body property from an AMQP message
+   * and returns the decoded message body. If it cannot decode the body then it returns the body
+   * as-is.
+   */
+  decode: (body: any) => any;
+}
+
+/**
+ * The default data transformer that will be used by the Azure SDK.
+ */
+export class DefaultDataTransformer implements DataTransformer {
+
+  /**
+   * A function that takes the body property from an EventData object
+   * and returns an encoded body (some form of AMQP type).
+   * @method
+   * @param {*} body The AMQP message body
+   * @return {DataSection} encodedBody - The encoded AMQP message body as an AMQP Data type
+   * (data section in rhea terms). Section object with following properties:
+   * - typecode: 117 (0x75)
+   * - content: The given AMQP message body as a Buffer.
+   * - multiple: true | undefined.
+   */
+  encode(body: any): any {
+    let result: any;
+    debug("The given message body that needs to be encoded is: ", body);
+    if (body !== undefined) {
+      if (isBuffer(body)) {
+        result = rhea.message.data_section(body);
+      } else if (typeof body === "string") {
+        result = rhea.message.data_section(Buffer.from(`"${body}"`, "utf8"));
+      } else {
+        // null, boolean, array, object, number should end up here
+        try {
+          const bodyStr = JSON.stringify(body);
+          result = rhea.message.data_section(Buffer.from(bodyStr, "utf8"));
+        } catch (err) {
+          const msg = `An error occurred while executing JSON.stringify() on the given body ` + body
+            + `${err ? err.stack : JSON.stringify(err)}`;
+          debug(msg);
+          throw new Error(msg);
+        }
+      }
+    } else {
+      // convert undefined to empty string.
+      result = rhea.message.data_section(Buffer.from("\"\"", "utf8"));
```

Comment:
why not just send an empty buffer? (`Buffer.alloc(0)`)

Reply:
Yes we can do that.
However while decoding we will get an error for doing JSON.parse('') (Since Buffer.alloc(0).toString() -> ''). Thus we will end up giving the empty Buffer back.

So the question is:
- Is it better to give an empty string or an empty Buffer object back when the user gave `undefined` as the message to be sent?

Reply:
maybe neither. i think coercing `undefined` to `null` makes more sense. that will stringify/parse correctly and null is closer to undefined than either of the other ones.

if i was picking between empty buffer and empty string, i'd pick empty buffer

Reply:
I see. I agree coercing to null is much better and will work well with other language SDKs since undefined is a JS thing.